services/grpc/internal/transformations/transformations.go

```diff
@@ -0,0 +1,91 @@
+package transformations
+
+const (
+	REPLACE = iota
+	TRANSCLUDE
+	DRAG
+	HIGHLIGHT
+)
+
+type FxnArgs struct {
+	FxnId        int
+	OffsetStart1 int
+	OffsetEnd1   int
+	OffsetStart2 int
+}
+
+// NOTE: How to get all functions to conform to signature: Call(string, string, *FxnArgs) (string, string, *FxnArgs) ?
+func Call(region0, src0 string, args *FxnArgs) (string, string, *FxnArgs) {
+	var region1, iregion string
+	var istart1, iend1, istart2 int
+
+	// NOTE: How to make functions extensible and discoverable?
+	switch args.FxnId {
+	case REPLACE:
+		region1, iregion, istart1, iend1 = Replace(region0, src0, args.OffsetStart1, args.OffsetEnd1)
+	case TRANSCLUDE:
+		region1, iregion, istart1, iend1 = Transclude(region0, src0, args.OffsetStart1, args.OffsetEnd1)
+	case DRAG:
+		region1, istart1, iend1, istart2 = Drag(region0, args.OffsetStart1, args.OffsetEnd1, args.OffsetStart2)
+	case HIGHLIGHT:
+		region1 = Highlight(region0, args.OffsetStart1, args.OffsetEnd1)
+	}
+
+	iargs := &FxnArgs{
+		FxnId:        args.FxnId,
+		OffsetStart1: istart1,
+		OffsetEnd1:   iend1,
+		OffsetStart2: istart2,
+	}
+
+	return region1, iregion, iargs
+}
+
+// Create the fist node in a document, with no source edges
+func CreateInitialNode(doc string) (*Node) {
+	// Create metadata, todo
+	var date, author string
+
+	node := &Node {
+		NodeId: GenerateUUID(),
+		CheckpointDoc: doc,
+		DocIsPopulated: true,
+	}
+
+	// Create a new scope for the new document
+	scope := &Scope {
+		ScopeId: GenerateUUID(),
+		Nodes: [node.NodeId]
+	}
+
+	node.scopes = append(node.scopes, scope)
+	return node
+}
+
+// Take in some source node, create an edge and a target node
+func ApplyNewTransformation(sourceNode *Node, args *FxnArgs, region0 string) (targetNode *Node, edge *Edge) {
+	// Create metadata, todo
+	var date, author string
+	
+	var sourceDoc string = AssembleDocForNode(*Node)
+	var targetDoc, inverseDoc string, invArgs *FxnArgs
+	targetDoc, inverseDoc, invArgs = Call(region0, sourceDoc, args)
+	
+	targetNode := &Node {
+		NodeId: GenerateUUID(),
+		DocIsPopulated: false,
+		Scopes: sourceNode.Scopes
+	}
+
+	edge := &Edge {
+		EdgeId: GenerateUUID(),
+		Args: args,
+		InverseArgs:invArgs,
+
+		SourceNodeId: sourceNode.NodeId,
+		TargetNodeId: targetNode.NodeId
+	}
```

Comment:
Hrm where did the constructor go? https://github.com/sourcegraph-xyz/sg/pull/5/files#diff-bcd403d2b86b5db40576a49ba6e877befe9d5485b8b96c4bafaec99575b73cd3R58 

I'm starting to think that when you rebased my changes were thrown out.

Reply:
Fixed, the changes are correct in the public-api package, just need to delete these old files